### PR TITLE
Improve the Flow type generator

### PR DESCRIPTION
**Pull request by keyz**
First of all, thank you for creating this library.

I made some changes to the Flow type generator:
1. Use `@flow strict`. This allows `@flow strict` projects to consume the type.
2. Use exact types (`{| ... |}`) by default. Now if you add an unknown property, Flow will warn. (See `__tests__/__snapshots__/flow.ts.snap`)
3. Use type spread (`...`) instead of intersection types (`&`). Intersection types are weird. We generally prefer using exact types and type spread. https://flow.org/en/docs/types/objects/#toc-exact-object-types
4. Lift all "fallbackable" types (e.g. `string | string[]`) to the top of the file. Unfortunately there's a bug in Flow (facebook/flow#7298) related to type spread and union types. An easy way to fix it is to lift the duplicated union types to type aliases.
5. Created a fixture and a snapshot for expected errors.

<details><summary>(here's the error if we don't lift the types)</summary>

```
Error ----------------------------------------------------------------------------------------------- typecheck.js:23:12

Could not decide which case to select. Since case 1 [1] may work but if it doesn't case 2 [2] looks promising too. To
fix add a type annotation to inferred union of array element types (alternatively, provide an annotation to summarize
the array element type) [3].

   typecheck.js:23:12
     23|   opacity: [1],
                    ^^^ [3]

References:
   index.js.flow:1798:13
   1798|   opacity?: GlobalsNumber | Array<GlobalsNumber>,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]
   index.js.flow:2355:13
   2355|   opacity?: GlobalsNumber | Array<GlobalsNumber>,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [2]


Error ----------------------------------------------------------------------------------------------- typecheck.js:25:15

Could not decide which case to select. Since case 1 [1] may work but if it doesn't case 2 [2] looks promising too. To
fix add a type annotation to inferred union of array element types (alternatively, provide an annotation to summarize
the array element type) [3].

   typecheck.js:25:15
     25|   fontWeight: ['normal'],
                       ^^^^^^^^^^ [3]

References:
   index.js.flow:1717:16
   1717|   fontWeight?: FontWeightProperty | Array<FontWeightProperty>,
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]
   index.js.flow:2344:16
   2344|   fontWeight?: FontWeightProperty | Array<FontWeightProperty>,
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [2]



Found 2 errors
```

</details>

### CODE REVIEW

**Comment by keyz** [comment]
friendly ping @frenic

**Comment by frenic** [comment]
Waw, I love this! Thank you! ❤️ 

First of, I'm not using Flow. So I'm hoping someone else with more Flow knowledge could do a review too. Maybe @rtsao?

Is if there's a way to use missing vendor specific properties or custom properties concerning it's using exact types by default? How would you solve this?

```js
const css: CSS.Properties<*> = {
  '--my-color': 'black',
};
```

This change should also be treated as a major change if I understand it corretly?

**Comment by TrySound** [comment]
@frenic If it's really required
```js
const css: { ...CSS.Properties<*> } = {
  '--my-color': 'black',
};
```

**Comment by TrySound** [comment]
Also this is equal to just `string` and makes all this types useless for value validating.
https://github.com/frenic/csstype/blob/master/index.js.flow#L4012
```js
export type AlignContentProperty = Globals | ContentDistribution | ContentPosition | "baseline" | "normal" | string;
```

**Comment by frenic** [comment]
> Also this is equal to just `string` and makes all this types useless for value validating.
> https://github.com/frenic/csstype/blob/master/index.js.flow#L4012
> 
> ```js
> export type AlignContentProperty = Globals | ContentDistribution | ContentPosition | "baseline" | "normal" | string;
> ```

It's useless for TypeScript too and will most likely forever be useless for validation. But we're hoping that one day that tools like VSCode will provide autocompletion using these literals. There's [an issue](https://github.com/frenic/csstype/issues/42) regarding this already. They are kept because there hasn't been any reason to remove them yet. No major performance benefits or anything like that since last time I checked. But the case could be different for Flow?

**Comment by keyz** [comment]
@frenic 

> Is if there's a way to use missing vendor specific properties or custom properties concerning it's using exact types by default? How would you solve this?

You can change it to an inexact type as @TrySound mentioned, but I think it'd be better to define the custom properties in userland. Something like:
```jsx
type ColorVariables = {|
  '--my-color': string,
  '--my-color-2': string,
|};

function getSomeStyle(): {|...CSS.Properties<*>, ...$Shape<ColorVariables>|} {
  ...
}
```

The nice thing about `$Shape<...>` is that it marks the fields optional, but still keeps the type exact.

> This change should also be treated as a major change if I understand it correctly?

Yeah, this would be a breaking change.

> Also this is equal to just string and makes all this types useless for value validating

I actually didn't notice that `string` is included in the types! Yeah I think it'd be nice to drop them, but that sounds like a bigger change.

File: src/output.ts [hunk 1]
```diff
@@ -8,44 +8,73 @@ export default () => ({
...
+      item.extends.map(extend => extend.name + stringifyGenerics(extend.generics, true)),
+    );
+
     interfacesOutput += 'export type ';
     interfacesOutput += item.name + stringifyGenerics(item.generics);
     interfacesOutput += ' = ' + extendList;
 
     if (item.properties.length > 0) {
       if (extendList) {
+        // ?? Is this actually getting used?
```
Comment (keyz) [inline src/output.ts]: Btw, @frantic is this codepath actually gonna get hit? I wonder this is a leftover from a refactor or something

**Comment by frenic** [comment]
> I actually didn't notice that `string` is included in the types! Yeah I think it'd be nice to drop them, but that sounds like a bigger change.

`string` is included where properties can have combined values like `display: block flow;` and experiments with combining literals is in the pipeline to get rid of `string` on some properties. I would be nice to see some kind of performance test results with removed string literals in Flow before considering getting rid of them permanently. Here's [a branch](https://github.com/frenic/csstype/tree/remove-unused-string-literals) with removed unnecessary string literals and completely resolved data types if someone would like to help out.

**Comment by keyz** [comment]
Thanks @frenic. I think the potential gains of dropping the `string` is more on the type-safety side. I don't know that much about Flow internals but I guess having `string` should actually make it more performant (since there're fewer "branches" to check).

**Comment by frenic** [comment]
I did a quick check on a huge Typescript application using CSS-in-JS with csstype and there's not even a tiny performance difference when type checking between the latest release of csstype and the one with removed string literals. Typescript simplifies `"foo" | "bar" | string`  early in the type checking process to just `string` so I guess that's why it doesn't affect performance. Hopefully it's the same for Flow. Let's keep that as is for now.

File: src/output.ts (hunk 1 above)
Comment (frenic) [reply src/output.ts]: No, it won't hit at the moment. But it wouldn't hurt if interfaces could extend other interfaces and have properties in the same time. I thought about something like this. What do you think?

```diff
@@ -14,7 +14,7 @@ function combineFlowExactTypes(input: string[]): string {
   } else if (input.length === 1) {
     return input[0];
   } else {
-    return '{|' + input.map(type => `...${type}`).join(`,${EOL}`) + '|}';
+    return input.map(type => `...${type}`).join(`,${EOL}`);
   }
 }

@@ -32,22 +32,14 @@ function flow() {
       interfacesOutput += EOL + EOL;
     }

-    const extendList = combineFlowExactTypes(
-      item.extends.map(extend => extend.name + stringifyGenerics(extend.generics, true)),
-    );
-
     interfacesOutput += 'export type ';
     interfacesOutput += item.name + stringifyGenerics(item.generics);
-    interfacesOutput += ' = ' + extendList;
+    interfacesOutput += ' = {|' + EOL;
+    interfacesOutput += combineFlowExactTypes(
+      item.extends.map(extend => extend.name + stringifyGenerics(extend.generics, true)),
+    );

     if (item.properties.length > 0) {
-      if (extendList) {
-        // ?? Is this actually getting used?
-        interfacesOutput += ' & ';
-      }
-
-      interfacesOutput += '{|' + EOL;
-
       for (const property of item.properties) {
         if (isAliasProperty(property)) {
           const generics = stringifyGenerics(property.generics, true);
@@ -73,9 +65,9 @@ function flow() {

         interfacesOutput += EOL;
       }
-
-      interfacesOutput += '|}';
     }
+
+    interfacesOutput += '|}';
```

File: src/output.ts [hunk 2]
```diff
@@ -66,7 +95,21 @@ function flow() {
...
 
-  return `// @flow ${EOL + interfacesOutput + EOL + EOL + declarationsOutput + EOL}`;
+  const fallbacksOutput = [...fallbackSet]
+    .map(name => {
+      return `type ${getNameForFallbackable(name)} = ${name} | Array<${name}>;`;
+    })
+    .join(EOL);
+
+  return `// @flow strict ${EOL +
+    fallbacksOutput +
```
Comment (frenic) [inline src/output.ts]: A comment why these fallback types are added here would be nice 👍

File: __tests__/__snapshots__/flow.ts.snap [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+References:
+   __fixtures__/flow-errors-fixture.js:4:12
+      4| const css: CSS.Properties<*> = {
+                    ^^^^^^^^^^^^^^^^^ [1]
+   __fixtures__/flow-errors-fixture.js:5:13
+      5|   flexGrow: 'invalid',
+                     ^^^^^^^^^ [3]
+   index.js.flow:4529:22
+   4529| type GlobalsNumber = Globals | number;
```
Comment (frenic) [inline __tests__/__snapshots__/flow.ts.snap]: These tests will fail every time these rows have changed. It needs to be addressed somehow.

File: __tests__/flow.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const path = require('path');
+const { spawnSync } = require('child_process');
```
Comment (frenic) [inline __tests__/flow.ts]: Please use import statement

File: __tests__/flow.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+test('it detects errors', () => {
+  const flowBin = path.resolve(__dirname, '../node_modules', '.bin', 'flow');
```
Comment (frenic) [inline __tests__/flow.ts]: The `.cmd` file muse be executed for Windows users.

```suggestion
  const flowBin = path.resolve(__dirname, '../node_modules', '.bin', process.platform === 'win32' ? 'flow.cmd' : 'flow',);
```

File: __tests__/__snapshots__/flow.ts.snap (hunk 3 above)
Comment (frenic) [reply __tests__/__snapshots__/flow.ts.snap]: Maybe it's better to request a JSON output using the `--json` flag and do tests against parts of the result of that instead?

Comment (keyz) [reply __tests__/__snapshots__/flow.ts.snap]: I think it's fine to have them fail when the rows are changed. Updating the snapshot is easy; having it in diff reviews serves as a sanity test for future changes.

File: src/output.ts (hunk 1 above)
Comment (keyz) [reply src/output.ts]: This seems to be more complicated than I expected. I added a TODO for now.

**Comment by frenic** [comment]
Thank you for this 👏 I'm merging this into separate branch as a part for v3.

**Pull request merged**